When scenario samplers are executing concurrently DB persistence fails

**Description:**

If Test scenario samplers are executing parallel the DB updating fails with the following error.

```
ERROR {} Detected problem in Listener. java.lang.RuntimeException: Error occurred when persisting test case.
	at org.wso2.testgrid.automation.executor.JMeterResultCollector.sampleOccurred(JMeterResultCollector.java:69)
	at org.apache.jmeter.threads.ListenerNotifier.notifyListeners(ListenerNotifier.java:65)
	at org.apache.jmeter.threads.JMeterThread.notifyListeners(JMeterThread.java:891)
	at org.apache.jmeter.threads.JMeterThread.executeSamplePackage(JMeterThread.java:528)
	at org.apache.jmeter.threads.JMeterThread.processSampler(JMeterThread.java:424)
	at org.apache.jmeter.threads.JMeterThread.run(JMeterThread.java:255)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.wso2.testgrid.dao.TestGridDAOException: Error occurred when persisting entity in database.
	at org.wso2.testgrid.dao.repository.AbstractRepository.persist(AbstractRepository.java:73)
	at org.wso2.testgrid.dao.repository.TestCaseRepository.persist(TestCaseRepository.java:53)
	at org.wso2.testgrid.dao.uow.TestCaseUOW.persistTestCase(TestCaseUOW.java:69)
	at org.wso2.testgrid.automation.executor.JMeterResultCollector.sampleOccurred(JMeterResultCollector.java:67)
	... 6 more
Caused by: java.lang.IllegalStateException: 
Exception Description: Transaction is currently active
	at org.eclipse.persistence.internal.jpa.transaction.EntityTransactionImpl.begin(EntityTransactionImpl.java:45)
	at org.wso2.testgrid.dao.repository.AbstractRepository.persist(AbstractRepository.java:65)
	... 9 more
```


**Suggested Labels:**
Bug

--- a/automation/src/main/java/org/wso2/testgrid/automation/executor/JMeterExecutor.java
+++ b/automation/src/main/java/org/wso2/testgrid/automation/executor/JMeterExecutor.java
@@ -20,7 +20,6 @@ package org.wso2.testgrid.automation.executor;
 
 import org.apache.commons.io.FileUtils;
 import org.apache.jmeter.engine.StandardJMeterEngine;
-import org.apache.jmeter.reporters.ResultCollector;
 import org.apache.jmeter.reporters.Summariser;
 import org.apache.jmeter.save.SaveService;
 import org.apache.jmeter.util.JMeterUtils;
@@ -31,8 +30,11 @@ import org.wso2.testgrid.automation.TestAutomationException;
 import org.wso2.testgrid.common.Deployment;
 import org.wso2.testgrid.common.Host;
 import org.wso2.testgrid.common.Port;
+import org.wso2.testgrid.common.TestCase;
 import org.wso2.testgrid.common.TestScenario;
 import org.wso2.testgrid.common.util.StringUtil;
+import org.wso2.testgrid.dao.TestGridDAOException;
+import org.wso2.testgrid.dao.uow.TestCaseUOW;
 
 import java.io.BufferedWriter;
 import java.io.File;
@@ -92,7 +94,7 @@ public class JMeterExecutor extends TestExecutor {
             throw new TestAutomationException(StringUtil.concatStrings("Script file ", script, " cannot be located."));
         }
 
-        ResultCollector resultCollector = new JMeterResultCollector(summariser, testScenario);
+        JMeterResultCollector resultCollector = new JMeterResultCollector(summariser, testScenario);
 
         if (testPlanTree.getArray().length == 0) {
             throw new TestAutomationException("JMeter test plan is empty.");
@@ -103,6 +105,18 @@ public class JMeterExecutor extends TestExecutor {
         jMeterEngine.configure(testPlanTree);
         jMeterEngine.run();
         jMeterEngine.exit();
+
+        //Persist all test cases for scenario
+        TestCaseUOW testCaseUOW = new TestCaseUOW();
+        for (TestCase testCase: resultCollector.getTestCases()) {
+            try {
+                testCaseUOW.persistTestCase(testCase);
+            } catch (TestGridDAOException e) {
+                throw new TestAutomationException(StringUtil.concatStrings("Error while persisting test case ",
+                        testCase.getName(), "of scenario ", testCase.getTestScenario().getName(), e));
+            }
+        }
+
         //delete temp file
         boolean delete = new File(script).delete();
         if (!delete) {

--- a/automation/src/main/java/org/wso2/testgrid/automation/executor/JMeterResultCollector.java
+++ b/automation/src/main/java/org/wso2/testgrid/automation/executor/JMeterResultCollector.java
@@ -80,7 +80,7 @@ public class JMeterResultCollector extends ResultCollector {
             testCase.setTestScenario(testScenario);
             testCase.setSuccess(result.isSuccessful());
             testCase.setFailureMessage(failureMessage);
-            testCases.add(testCase);
+            testScenario.addTestCase(testCase);
             jmeterResultLogger.info(StringUtil.concatStrings("Test case :", result.getSampleLabel(),
                     " failed for scenario: ", testScenario.getName(), "\n", failureMessage, "\", \"Sampler Data\": \"",
                     result.getSamplerData().replaceAll("\"", "\\\""), "\"}"));

--- a/automation/src/main/java/org/wso2/testgrid/automation/executor/JMeterResultCollector.java
+++ b/automation/src/main/java/org/wso2/testgrid/automation/executor/JMeterResultCollector.java
@@ -23,10 +23,12 @@ import org.apache.jmeter.samplers.SampleEvent;
 import org.apache.jmeter.samplers.SampleResult;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.wso2.testgrid.common.TestCase;
 import org.wso2.testgrid.common.TestScenario;
 import org.wso2.testgrid.common.util.StringUtil;
-import org.wso2.testgrid.dao.TestGridDAOException;
-import org.wso2.testgrid.dao.uow.TestCaseUOW;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
 
 /**
  * This class is responsible for collecting JMeter test results.
@@ -41,6 +43,7 @@ public class JMeterResultCollector extends ResultCollector {
 
     private TestScenario testScenario;
     private static final int FAILURE_MESSAGE_LIMIT = 5000;
+    private List<TestCase> testCases = new CopyOnWriteArrayList<>();
 
     /**
      * Constructs an instance of {@link JMeterResultCollector}.
@@ -55,8 +58,6 @@ public class JMeterResultCollector extends ResultCollector {
 
     @Override
     public void sampleOccurred(SampleEvent sampleEvent) {
-        try {
-            TestCaseUOW testCaseUOW = new TestCaseUOW();
             super.sampleOccurred(sampleEvent);
             SampleResult result = sampleEvent.getResult();
             String samplerData = result.getSamplerData();
@@ -73,13 +74,20 @@ public class JMeterResultCollector extends ResultCollector {
                                             result.getResponseMessage().replaceAll("\"", "\\\""),
                                             "\", \"Sampler Data\": \"",
                                             samplerData.replaceAll("\"", "\\\""), "\"}");
-            // Persist result to the database
-            testCaseUOW.persistTestCase(result.getSampleLabel(), testScenario, result.isSuccessful(), failureMessage);
+            // Add result to concurrent list
+            TestCase testCase = new TestCase();
+            testCase.setName(result.getSampleLabel());
+            testCase.setTestScenario(testScenario);
+            testCase.setSuccess(result.isSuccessful());
+            testCase.setFailureMessage(failureMessage);
+            testCases.add(testCase);
             jmeterResultLogger.info(StringUtil.concatStrings("Test case :", result.getSampleLabel(),
                     " failed for scenario: ", testScenario.getName(), "\n", failureMessage, "\", \"Sampler Data\": \"",
                     result.getSamplerData().replaceAll("\"", "\\\""), "\"}"));
-        } catch (TestGridDAOException e) {
-            throw new RuntimeException(StringUtil.concatStrings("Error occurred when persisting test case."), e);
-        }
+
+    }
+
+    public List<TestCase> getTestCases() {
+        return testCases;
     }
 }

--- a/automation/src/main/java/org/wso2/testgrid/automation/executor/JMeterResultCollector.java
+++ b/automation/src/main/java/org/wso2/testgrid/automation/executor/JMeterResultCollector.java
@@ -87,6 +87,13 @@ public class JMeterResultCollector extends ResultCollector {
 
     }
 
+    /**
+     * Gets the list of test cases pertaining to the scenario.
+     * The test cases are added to the into the list concurrently once it enters
+     * {@link #sampleOccurred(SampleEvent)} and the test execution is complete.
+     *
+     * @return the concurrent list of test cases
+     */
     public List<TestCase> getTestCases() {
         return testCases;
     }

--- a/dao/src/main/java/org/wso2/testgrid/dao/uow/TestCaseUOW.java
+++ b/dao/src/main/java/org/wso2/testgrid/dao/uow/TestCaseUOW.java
@@ -47,25 +47,10 @@ public class TestCaseUOW {
     /**
      * Persists the {@link TestCase} instance for the given params.
      *
-     * @param testName        name of the test case
-     * @param testScenario    {@link TestScenario} instance associated with test case
-     * @param isSuccess       status of the TestCase
-     * @param responseMessage response message of the test case
+     * @param testCase        test case to persist
      * @throws TestGridDAOException thrown when error on persisting
      */
-    public TestCase persistTestCase(String testName, TestScenario testScenario, boolean isSuccess,
-                                    String responseMessage) throws TestGridDAOException {
-        // Create test case instance
-        TestCase testCase = new TestCase();
-        testCase.setName(testName);
-        testCase.setSuccess(isSuccess);
-        testCase.setTestScenario(testScenario);
-        testCase.setFailureMessage(responseMessage);
-
-        // Set test scenario -> test case link
-        testScenario.addTestCase(testCase);
-
-        // Persist test case
+    public TestCase persistTestCase(TestCase testCase) throws TestGridDAOException {
         return testCaseRepository.persist(testCase);
     }
 

